### PR TITLE
fix: replace gzip with zstd for faster backup packaging and to prevent backup errors

### PR DESCRIPTION
**Pull request by Xm798**
### Background:
Our previous packaging process with gzip was not only time-consuming but also susceptible to generating warning messages if files were modified during the lengthy packaging period. Such warnings could be mistakenly interpreted by the backupmanager script as a failure in the backup process.
For example:
```
palworld-dedicated-server  | > RCON: Broadcasted: Creating-backup
palworld-dedicated-server  | tar: Saved/SaveGames/0/xxxxx/backup/world: file changed as we read it
palworld-dedicated-server  | tar: Saved/SaveGames/0/xxxxx: file changed as we read it
palworld-dedicated-server  | [2024-04-04 14:00:12] [LOG] RCON executed the command. broadcast Backup-failed
palworld-dedicated-server  | > RCON: Broadcasted: Backup-failed
```

### Solution:
By switching to zstd, we're speeding up our packaging process big time. This means there's less chance for files to get changed while we're packaging, which cuts down on those annoying warning messages. Before, these warnings could make us think our backup failed when it actually didn't. On my machine, packing a 200MB world archive used to take 12 seconds with gzip, but now with zstd, it's down to just about 0.8 seconds. That's a huge boost in how quickly and reliably we can manage our backups.

### CODE REVIEW

**Comment by Xm798** [comment]
If there are specific reasons why we must stick with gz, then I suggest we add the `--warning=no-file-changed` and `--warning=no-file-removed` options to avoid the common warning messages we're likely to encounter. For more details, check out: https://www.gnu.org/software/tar/manual/html_section/warnings.html

**Comment by Callum027** [comment]
If `--warning=no-file-changed` is an algorithm-agnostic parameter, perhaps we should set it in addition to switching to zstd?

**Comment by Xm798** [comment]
> If `--warning=no-file-changed` is an algorithm-agnostic parameter, perhaps we should set it in addition to switching to zstd?

That's a good point. In my tests, zstd was so fast that I didn't encounter any file change warnings. But adding --warning=no-file-changed might indeed be a more robust approach. Thanks for the suggestion!

**Comment by Callum027** [comment]
Hey @Xm798, I've got a couple of other suggestions for this:

* I also meant to say, also add `--warning=no-file-removed` in addition to changed, regardless of whether we stick with gzip or switch to zstd.
* In the new version, Palworld now has its own built-in backup mechanism, which gets saved to the `backup` directory. This folder is being included with the container backup, exponentially increasing the size of the backups. As double backups are not really useful, an you also add `--exclude=backup` to make sure those files are not included into the tarball?

@jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip:

* https://jothiprasath.com/blog/gzip-vs-zstd
* https://manishrjain.com/compression-algo-moving-data

File: scripts/backupmanager.sh [hunk 1]
```diff
@@ -144,7 +144,7 @@ function create_backup() {
     sleep 15
     rconcli broadcast "$(get_time) Creating backup..."
 
-    if ! tar cfz "${LOCAL_BACKUP_PATH}/${backup_file_name}" -C "${LOCAL_GAME_PATH}/" "Saved" ; then
+    if ! tar --ignore-failed-read --warning=no-file-changed --warning=no-file-removed -I zstd -cf "${LOCAL_BACKUP_PATH}/${backup_file_name}" -C "${LOCAL_GAME_PATH}/" "Saved" ; then
```
Comment (Callum027) [inline scripts/backupmanager.sh:147]: In what situation would `--ignore-failed-read` be relevant?

I'm inclined to make the backup fail in this case. Being unable to read files seems like a valid reason to not create a backup.

Comment (jammsen) [reply scripts/backupmanager.sh:147]: Looks to me like corrupt archive, when this goes through, im with Callum.

**Comment by jammsen** [comment]
> @jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip:
> 
> * https://jothiprasath.com/blog/gzip-vs-zstd
> * https://manishrjain.com/compression-algo-moving-data
> * https://community.centminmod.com/threads/round-3-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.17259

Yeah i knew i could google for it, but i wanted to see his data. Thats why i asked. He says its very much faster, but never shows it.

**Comment by Xm798** [comment]
> > @jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip: - 快速谷歌搜索会显示一些页面，证明 zstd 压缩（级别 3 左右）比 gzip 更快并且创建的文件更小：
> > 
> > * https://jothiprasath.com/blog/gzip-vs-zstd
> > * https://manishrjain.com/compression-algo-moving-data
> > * https://community.centminmod.com/threads/round-3-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.17259
> 
> Yeah i knew i could google for it, but i wanted to see his data. Thats why i asked. He says its very much faster, but never shows it.是的，我知道我可以用谷歌搜索它，但我想看看他的数据。这就是我问的原因。他说速度快得多，但从未表现出来。

I conducted another test on my Palworld server, and the data is approximately as follows:

```bash
╰─❯ time tar cfz backup_test.tgz Saved
tar: Saved/SaveGames/0/2EFA86D24BF543E7AFFD4C18D5FCE34E/backup/world: file changed as we read it
tar: Saved/SaveGames/0/2EFA86D24BF543E7AFFD4C18D5FCE34E/Level.sav: file changed as we read it
tar: Saved/SaveGames/0/2EFA86D24BF543E7AFFD4C18D5FCE34E: file changed as we read it
tar cfz backup_test.tgz Saved  14.26s user 0.93s system 101% cpu 14.939 total

╰─❯ time tar -I zstd -cf backup_test.tar.zst Saved
tar -I zstd -cf backup_test.tar.zst Saved  0.58s user 0.90s system 187% cpu 0.793 total
```

A 285MB Saved folder, when compressed with gzip, took about 15 seconds, and there was a warning from tar about files being modified. In contrast, zstd completed the compression in just about 0.8 seconds, successfully finishing the packaging.

image

**Comment by Xm798** [comment]
> > @jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip: - 快速谷歌搜索会显示一些页面，证明 zstd 压缩（级别 3 左右）比 gzip 更快并且创建的文件更小：
> > 
> > * https://jothiprasath.com/blog/gzip-vs-zstd
> > * https://manishrjain.com/compression-algo-moving-data
> > * https://community.centminmod.com/threads/round-3-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.17259
> 
> Yeah i knew i could google for it, but i wanted to see his data. Thats why i asked. He says its very much faster, but never shows it.是的，我知道我可以用谷歌搜索它，但我想看看他的数据。这就是我问的原因。他说速度快得多，但从未表现出来。

image


> Hey @Xm798, I've another suggestions for this.嘿，对此我还有另一个建议。
> 
> In the new version, Palworld now has its own built-in backup mechanism, which gets saved to the `backup` directory. This folder is being included with the container backup, [exponentially increasing the size of the backups](https://github.com/jammsen/docker-palworld-dedicated-server/issues/257). As double backups are not really useful, can you also add `--exclude=backup` to make sure those files are not included into the tarball?在新版本中，Palworld现在有自己的内置备份机制，该机制会保存到 `backup` 目录中。该文件夹包含在容器备份中，从而以指数方式增加备份的大小。由于双重备份并不是很有用，您是否还可以添加 `--exclude=backup` 以确保这些文件不包含在 tarball 中？
> 
> @jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip: - 快速谷歌搜索会显示一些页面，证明 zstd 压缩（级别 3 左右）比 gzip 更快并且创建的文件更小：
> 
> * https://jothiprasath.com/blog/gzip-vs-zstd
> * https://manishrjain.com/compression-algo-moving-data
> * https://community.centminmod.com/threads/round-3-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.17259

1. Based on the [GNU's tar documentation](https://www.gnu.org/software/tar/manual/html_section/warnings.html), I added the --no-file-changed option as "file-changed keyword applies only if used together with the ‘--ignore-failed-read’ option".

2. Upon reevaluating the current setup post your feedback, I noticed that indeed, following the recent update two days ago, Palworld automatically initiates world saves in the /Pal/Saved/SaveGames/0/xxxx/ path. This new feature correlates with the increased frequency of tar backup errors, likely because the backup folder is over 200M in size, slowing down the tar process while Palworld simultaneously modifies the files.

3. Consequently, considering the insights and the observed redundancy leading to space wastage, it might be wise to heed @Callum027 advice to exclude the Pal backup folder from our backups. The dual backup system is evidently consuming excess storage, which became painfully apparent when my hard drive quickly filled up today—truly a frustrating situation.

**Comment by Xm798** [comment]
> Hey @Xm798, I've another suggestions for this.
> 
> In the new version, Palworld now has its own built-in backup mechanism, which gets saved to the `backup` directory. This folder is being included with the container backup, [exponentially increasing the size of the backups](https://github.com/jammsen/docker-palworld-dedicated-server/issues/257). As double backups are not really useful, can you also add `--exclude=backup` to make sure those files are not included into the tarball?
> 
> @jammsen - Quick googling brings up a few pages that demonstrate that zstd compression, at level 3 or so, is both faster and creates smaller files than gzip:
> 
> * https://jothiprasath.com/blog/gzip-vs-zstd
> * https://manishrjain.com/compression-algo-moving-data
> * https://community.centminmod.com/threads/round-3-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.17259

1. Based on the [GNU's tar documentation](https://www.gnu.org/software/tar/manual/html_section/warnings.html), I added the --no-file-changed option as "file-changed keyword applies only if used together with the ‘--ignore-failed-read’ option".

2. Upon reevaluating the current setup post your feedback, I noticed that indeed, following the recent update two days ago, Palworld automatically initiates world saves of the /Pal/Saved/SaveGames/0/xxxx/ path. This new feature correlates with the increased frequency of tar backup errors, likely because the backup folder is over 200M in size (on my Pal server), slowing down the tar process while Palworld simultaneously modifies the files.

3. Consequently, considering the insights and the observed redundancy leading to space wastage, it might be wise to heed @Callum027 advice to exclude the Pal backup folder from our backups. The dual backup system is evidently consuming excess storage, which became painfully apparent when my hard drive quickly filled up today—truly a frustrating situation.

**Comment by Callum027** [comment]
> 1. Based on the [GNU's tar documentation](https://www.gnu.org/software/tar/manual/html_section/warnings.html), I added the --no-file-changed option as "file-changed keyword applies only if used together with the ‘--ignore-failed-read’ option".

This didn't make sense, so I checked `man tar` on my machines.

image

Turns out that only applies to `--warning=failed-read` (the website appears to have rendered the manual incorrectly). You don't need `--ignore-failed-read` set for the warnings you've disabled in this PR.

File: scripts/backupmanager.sh (hunk 1 above)
Comment (Callum027) [reply scripts/backupmanager.sh:147]: As I commented [here](https://github.com/jammsen/docker-palworld-dedicated-server/pull/254#issuecomment-2041036712), I don't believe this is necessary for the PR to work.

**Comment by Xm798** [comment]
noticed that PR#259 submitted a modification to exclude the backup directory, which is exactly what I intended to do later. After excluding the backups managed by Palworld, the packaging time has been reduced to an acceptable level again (about 900ms for me using gz), and there were no packaging errors. Therefore, this PR can be closed now. Thank you all for your help.

**Comment by jammsen** [comment]
@Xm798 - No thanks for your help and your contribution!

**Pull request closed without merging**